ddtrace/context.py

```diff
@@ -74,6 +88,63 @@ def _update_tags(self, span):
             span.meta.update(self._meta)
             span.metrics.update(self._metrics)
 
+    def _update_upstream_services(
+        self,
+        span,  # type: Span
+        sampling_priority,  # type: int
+        sampling_mechanism,  # type: SamplingMechanism
+        sample_rate=None,  # type: Optional[float]
+    ):
+        # type: (...) -> None
+        """
+        Update the ``_dd.p.upstream_services`` tag associated with this context.
+
+        Calling with the same values multiple times will not change anything.
+
+        Calling multiple times with new values will replace the previous entry with the new one.
+        (We will only have 1 entry in the tag for this context)
+        """
+
+        entry = "|".join(
+            [
+                # base64 encode binary representation of the service name
+                # DEV: remove padding
+                ensure_text(base64.b64encode(ensure_binary(span.service or "unnamed-python-service"))).strip("="),
+                str(sampling_priority),
+                str(sampling_mechanism.value),
+                # Float string rounded to 4 decimal places if set, otherwise an empty string
+                "{:0.4f}".format(sample_rate) if sample_rate is not None else "",
+            ]
+        )
+
+        # If the sampling decision hasn't changed, do not add an entry
+        # For example, if someone does `span.set_tag(MANUAL_KEEP_KEY)` multiple times in one trace
+        if self._last_upstream_service_entry == entry:
+            return
+
+        # Fetch any existing value (extracted from propagation headers, or we have already set ourselves)
+        existing = self._meta.get(UPSTREAM_SERVICES_KEY)
+
+        # We have an existing header value and a previously set upstream service entry by us
+        # DEV: We can have an existing value from another upstream service, it doesn't mean it was set by us
+        if existing and self._last_upstream_service_entry is not None:
+            # Remove the last entry if one exists
+            existing, _, old = existing.rpartition(";")
+            # No `;` found, we only had the one entry / our last entry
+            if not old:
+                self._meta[UPSTREAM_SERVICES_KEY] = entry
+            else:
+                self._meta[UPSTREAM_SERVICES_KEY] = ";".join([existing, entry])
+
+        # Existing upstream value, and none set by us yet
+        elif existing:
+            self._meta[UPSTREAM_SERVICES_KEY] = ";".join([existing, entry])
+        else:
+            self._meta[UPSTREAM_SERVICES_KEY] = entry
+
+        # Keep track of the last entry that we created/added
+        self._last_upstream_service_entry = entry
```

Comment:
Another approach here could be to keep track of the inherited/extracted `_dd.p.upstream_services` from context propagation separate from "current entry", that way this entire thing just ends up being ("ish"):

```python
self._upstream_service_entry = "|".join([ base_64, priority, mechanism, rate ])
self._meta[UPSTREAM_SERVICES_KEY] = ";".join([
    entry
    for entry in (self._parent_upstream_service_entry, self._upstream_service_entry)
    if entry
])
```